agent/pkg/osauth/auth.go

```diff
@@ -1,55 +1,145 @@
+	_ "github.com/GehirnInc/crypt/md5_crypt"
+	_ "github.com/GehirnInc/crypt/sha256_crypt"
+	_ "github.com/GehirnInc/crypt/sha512_crypt"
+	"github.com/sirupsen/logrus"
 )
 
 var DefaultShadowFilename = "/etc/shadow"
 
-func AuthUser(user, passwd string) bool {
-	cuser := C.CString(user)
-	defer C.free(unsafe.Pointer(cuser))
+type ShadowEntry struct {
+	Username    string
+	Password    string
+	Lastchanged int
+	Minimum     int
+	Maximum     int
+	Warn        int // The number of days before password is to expire that user is warned that his/her password must be changed
+	Inactive    int // The number of days after password expires that account is disabled
+	Expire      int
+}
 
-	cpasswd := C.CString(passwd)
-	defer C.free(unsafe.Pointer(cpasswd))
+func AuthUser(username, passwd string) bool {
+	shadowFile, err := os.Open(DefaultShadowFilename)
+	if err != nil {
+		// TODO: log error
+		logrus.Error("Could not open /etc/shadow")
+		return false
+	}
 
-	cfilename := C.CString(DefaultShadowFilename)
-	defer C.free(unsafe.Pointer(cfilename))
+	defer shadowFile.Close()
 
-	cmode := C.CString("r")
-	defer C.free(unsafe.Pointer(cmode))
+	entries, err := parseShadowReader(shadowFile)
+	if err != nil {
+		logrus.Println("Could not parse shadowfile %v", err)
+		return false
+	}
 
-	f := C.fopen(cfilename, cmode)
-	defer C.fclose(f)
+	if entry, ok := entries[username]; ok {
+		if entry.Password != "" {
+			crypt := crypt.NewFromHash(entry.Password)
+			if crypt == nil {
+				logrus.Error("Could not detect password crypto algorithm from shadow entry")
+				return false
+			}
 
-	var pwd *C.struct_spwd
-	for {
-		if pwd = C.fgetspent(f); pwd == nil {
-			return false
+			err := crypt.Verify(entry.Password, []byte(passwd))
+			return err == nil
 		}
+		logrus.Error("Password entry is empty")
+		return false
+	}
+
+	logrus.Warn("User not found")
+	return false
+}
 
-		if C.strcmp(cuser, pwd.sp_namp) == 0 {
+func parseShadowReader(r io.Reader) (map[string]ShadowEntry, error) {
+	lines := bufio.NewReader(r)
+	entries := make(map[string]ShadowEntry)
+	for {
+		line, _, err := lines.ReadLine()
+		if err != nil {
 			break
 		}
+
+		if len(line) == 0 || strings.HasPrefix(string(line), "#") {
+			continue
+		}
+
+		entry, err := parseShadowLine(string(line))
+		if err != nil {
+			return nil, err
+		}
+
+		entries[entry.Username] = entry
 	}
+	return entries, nil
+}
 
-	if pwd == nil {
-		return false
+func parseShadowLine(line string) (ShadowEntry, error) {
+	result := ShadowEntry{}
+	parts := strings.Split(strings.TrimSpace(line), ":")
+	if len(parts) != 9 {
+		return result, fmt.Errorf("Shadow line had wrong number of parts %d != 9", len(parts))
 	}
+	result.Username = strings.TrimSpace(parts[0])
+	result.Password = strings.TrimSpace(parts[1])
+	/*
+	   	lastchanged, err := strconv.Atoi(parts[2])
+	   	if err != nil {
+	   		return result, fmt.Errorf("Shadow line had badly formatted lastchanged %s", parts[2])
+	   	}
+	   	result.Lastchanged = lastchanged
```

Comment:
I think we ought to include this code so it becomes fully featured. The hassle of adding it back will be bigger than adding it now.

Reply:
Ok, added it back in, also made it a little nicer

Reply:
Cool, could you rebase the patches and squash this last one onto the original one? There is no reason for it to be done in a specific commit.

Reply:
ok, I can also squash the entire PR if that makes more sense ?

Reply:
No; small and logical commits are nicer for repo history.